ADAL/src/urlprotocol/ADURLSessionDemux.m

```diff
@@ -0,0 +1,234 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "ADURLSessionDemux.h"
+
+
+@interface ADURLSessionDemuxTaskInfo : NSObject
+
+- (instancetype)initWithTask:(NSURLSessionDataTask *)task delegate:(id<NSURLSessionDataDelegate>)delegate;
+
+@property (atomic, strong) NSURLSessionDataTask *task;
+@property (atomic, strong) id<NSURLSessionDataDelegate> delegate;
+@property (atomic, strong) NSThread *thread;
+
+- (void)performBlock:(dispatch_block_t)block;
+
+- (void)invalidate;
+
+@end
+
+@implementation ADURLSessionDemuxTaskInfo
+
+- (instancetype)initWithTask:(NSURLSessionDataTask *)task delegate:(id<NSURLSessionDataDelegate>)delegate
+{
+    self = [super init];
+    if (self != nil)
+    {
+        self->_task = task;
+        self->_delegate = delegate;
+        self->_thread = [NSThread currentThread];
+    }
+    return self;
+}
+
+- (void)performBlock:(dispatch_block_t)block
+{
+    [self performSelector:@selector(performBlockOnClientThread:)
+                 onThread:self.thread
+               withObject:[block copy]
+            waitUntilDone:NO];
+}
+
+- (void)performBlockOnClientThread:(dispatch_block_t)block
+{
+    block();
+}
+
+- (void)invalidate
+{
+    self.delegate = nil;
+    self.thread = nil;
+}
+
+@end
+
+
+@interface ADURLSessionDemux() <NSURLSessionDataDelegate>
+
+@property (atomic, strong, readonly) NSMutableDictionary *taskInfoByTaskID;
+@property (atomic, strong, readonly) NSOperationQueue *sessionDelegateQueue;
+
+@end
+
+
+@implementation ADURLSessionDemux
+
+- (instancetype)initWithConfiguration:(NSURLSessionConfiguration *)configuration
+{
+    self = [super init];
+    if (self)
+    {
+        self->_configuration = [configuration copy];
+        
+        self->_taskInfoByTaskID = [[NSMutableDictionary alloc] init];
+        
+        self->_sessionDelegateQueue = [[NSOperationQueue alloc] init];
+        [self->_sessionDelegateQueue setMaxConcurrentOperationCount:1];
+        [self->_sessionDelegateQueue setName:@"ADURLSessionDemux"];
+
+        self->_session = [NSURLSession sessionWithConfiguration:self->_configuration delegate:self delegateQueue:self->_sessionDelegateQueue];
+        self->_session.sessionDescription = @"ADURLSessionDemux";
+    }
+    
+    return self;
+}
+
+- (NSURLSessionDataTask *)dataTaskWithRequest:(NSURLRequest *)request delegate:(id<NSURLSessionDataDelegate>)delegate
+{
+    NSURLSessionDataTask *task;
+    ADURLSessionDemuxTaskInfo *taskInfo;
+    
+    task = [self.session dataTaskWithRequest:request];
+    taskInfo = [[ADURLSessionDemuxTaskInfo alloc] initWithTask:task
```

Comment:
I think the @synchronized is being used to protect the taskInfoByTaskID mutable dictionary here, I'm not seeing the reason why the other lines would need to be protected.